cmd/agent/daemon/enrichment/enrichers.go

```diff
@@ -0,0 +1,149 @@
+		panic(err)
+	}
+
+	return &fileHashEnricher{
+		log:                      log,
+		containerForCgroupGetter: containerForCgroupGetter,
+		procFS:                   procFS,
+		cache:                    cache,
+	}
+}
+
+type fileHashEnricher struct {
+	log                      *logging.Logger
+	containerForCgroupGetter ContainerForCgroupGetter
+	procFS                   fs.StatFS
+	cache                    *lru.Cache[fileHashCacheKey, []byte]
+}
+
+func (enricher *fileHashEnricher) Enrich(ctx context.Context, e *castpb.Event) {
+	exec := e.GetExec()
+	if exec == nil || exec.Path == "" {
+		return
+	}
+
+	cont, found, err := enricher.containerForCgroupGetter(e.CgroupId)
+	// if we do not have the container cached, we cannot easily accesss the FS and hence simply skip
+	if err != nil || !found {
+		return
+	}
+
+	for _, pid := range cont.PIDs {
+		sha, err := enricher.calcFileHashForPID(cont, pid, exec.Path)
+		// We search for the first PID we can successfully calculate a filehash for.
+		if err != nil {
+			if errors.Is(err, ErrFileDoesNotExist) {
+				// If the wanted file does not exist in the PID mount namespace, it will also not exist in the mounts of the other.
+				// We can hence simply return, as we will not find the wanted file.
+				return
+			}
+
+			continue
+		}
+
+		if exec.Meta == nil {
+			exec.Meta = &castpb.ExecMetadata{}
+		}
+
+		exec.Meta.HashSha256 = sha
+		return
+	}
+}
+
+func (enricher *fileHashEnricher) calcFileHashForPID(cont *containers.Container, pid uint32, execPath string) ([]byte, error) {
+	pidString := strconv.FormatInt(int64(pid), 10)
+
+	_, err := enricher.procFS.Stat(pidString)
+	if err != nil {
+		// If the /proc/<pid> folder doesn't exist, there is nothing we can do.
+		return nil, ErrProcFolderDoesNotExist
+	}
+
+	path := filepath.Join(pidString, "root", execPath)
+	info, err := enricher.procFS.Stat(path)
+	if err != nil {
+		// If the wanted file does not exist inside the mount namespace, there is also nothing we can do.
+		return nil, ErrFileDoesNotExist
+	}
+
+	key := enricher.buildCacheKey(cont, info)
+	hash, found := enricher.checkCache(key)
+	if found {
+		return hash, nil
+	}
+
+	f, err := enricher.procFS.Open(path)
+	if err != nil {
+		return nil, ErrFileDoesNotExist
+	}
+	defer f.Close()
+
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return nil, err
+	}
+
+	hash = h.Sum(nil)
+	enricher.cacheHash(key, hash)
+
+	return hash, nil
+}
+
+var (
+	ErrCannotGetInode         = errors.New("cannot get inode for path")
+	ErrProcFolderDoesNotExist = errors.New("/proc/<pid> folder does not exist")
+	ErrFileDoesNotExist       = errors.New("wanted file does not exist")
+)
+
+func (enricher *fileHashEnricher) buildCacheKey(cont *containers.Container, info fs.FileInfo) fileHashCacheKey {
+	stat, ok := info.Sys().(*syscall.Stat_t)
+	if !ok {
+		return ""
+	}
+
+	return fileHashCacheKey(fmt.Sprintf("%s:%d", cont.Cgroup.ContainerID, stat.Ino))
```

Comment:
Would it be possible to use container's image digest here? Like `fmt.Sprintf("%s:%d", cont.ImageDigest, stat.Ino)`

Reply:
I had a look at how tracee implements it and they are actually just using the image name it seems ([source](https://github.com/aquasecurity/tracee/blob/6447ca2ff75e744c9825ac79ea3ba3a731d2f01f/pkg/containers/runtime/containerd.go#L67)). 

The problem with containerd is, that the `image` field on the container struct we get back is also not the digest. We would need to resolve it via the container registry.

Image name could yield false results though, as the underlying image may have changed. I am not 100% sure how to continue here tbh.